[Merged by Bors] - Add Holesky

paulhauner opened the pull request:
## Issue Addressed

NA

## Proposed Changes

Add the Holesky network config as per https://github.com/eth-clients/holesky/tree/36e4ff2d5138dcb2eb614f0f60fdb060b2adc1e2/custom_config_data.

Since the genesis state is ~190MB, I've opted to *not* include it in the binary and instead download it at runtime (see #4564 for context). To download this file we have:

- A hard-coded URL for a SigP-hosted S3 bucket with the Holesky genesis state. Assuming this download works correctly, users will be none the wiser that the state wasn't included in the binary (apart from some additional logs)
- If the user provides a `--checkpoint-sync-url` flag, then LH will download the genesis state from that server rather than our S3 bucket.
- If the user provides a `--genesis-state-url` flag, then LH will download the genesis state from that server regardless of the S3 bucket or `--checkpoint-sync-url` flag.
- Whenever a genesis state is downloaded it is checked against a checksum baked into the binary.
- A genesis state will never be downloaded if it's already included in the binary.
- There is a `--genesis-state-url-timeout` flag to tweak the timeout for downloading the genesis state file.

## Log Output

Example of log output when a state is downloaded:

```bash
Aug 23 05:40:13.424 INFO Logging to file                         path: "/Users/paul/.lighthouse/holesky/beacon/logs/beacon.log"
Aug 23 05:40:13.425 INFO Lighthouse started                      version: Lighthouse/v4.3.0-bd9931f+
Aug 23 05:40:13.425 INFO Configured for network                  name: holesky
Aug 23 05:40:13.426 INFO Data directory initialised              datadir: /Users/paul/.lighthouse/holesky
Aug 23 05:40:13.427 INFO Deposit contract                        address: 0x4242424242424242424242424242424242424242, deploy_block: 0
Aug 23 05:40:13.427 INFO Downloading genesis state               info: this may take some time on testnets with large validator counts, timeout: 60s, server: https://sigp-public-genesis-states.s3.ap-southeast-2.amazonaws.com/
Aug 23 05:40:29.895 INFO Starting from known genesis state       service: beacon
```

Example of log output when there are no URLs specified:

```
Aug 23 06:29:51.645 INFO Logging to file                         path: "/Users/paul/.lighthouse/goerli/beacon/logs/beacon.log"
Aug 23 06:29:51.646 INFO Lighthouse started                      version: Lighthouse/v4.3.0-666a39c+
Aug 23 06:29:51.646 INFO Configured for network                  name: goerli
Aug 23 06:29:51.647 INFO Data directory initialised              datadir: /Users/paul/.lighthouse/goerli
Aug 23 06:29:51.647 INFO Deposit contract                        address: 0xff50ed3d0ec03ac01d4c79aad74928bff48a7b2b, deploy_block: 4367322
The genesis state is not present in the binary and there are no known download URLs. Please use --checkpoint-sync-url or --genesis-state-url.
```

## Additional Info

I tested the `--genesis-state-url` flag with all 9 Goerli checkpoint sync servers on https://eth-clients.github.io/checkpoint-sync-endpoints/ and they all worked 🎉 

My IDE eagerly formatted some `Cargo.toml`. I've disabled it but I don't see the value in spending time reverting the changes that are already there.

I also added the `GenesisStateBytes` enum to avoid an unnecessary clone on the genesis state bytes baked into the binary. This is not a huge deal on Mainnet, but will become more relevant when testing with big genesis states.

When we do a fresh checkpoint sync we're downloading the genesis state to check the `genesis_validators_root` against the finalised state we receive. This is not *entirely* pointless, since we verify the checksum when we download the genesis state so we are actually guaranteeing that the finalised state is on the same network. There might be a smarter/less-download-y way to go about this, but I've run out of cycles to figure that out. Perhaps we can grab it in the next release?

paulhauner committed:
Add holesky via Github download

paulhauner committed:
Add checksum

paulhauner committed:
Simplify checksum deserialization

paulhauner committed:
Support using the checkpoint sync URL

paulhauner committed:
Always assume checkpointz server

paulhauner committed:
Tidy, fix bugs

paulhauner committed:
Pass logger into download function

paulhauner committed:
Add timeout flag

paulhauner committed:
Fix preferences for state url

paulhauner committed:
Fix clippy lints

paulhauner committed:
Fix failing test and error message

paulhauner committed:
Load genesis state on demand

paulhauner committed:
Avoid cloning genesis state bytes

paulhauner committed:
Only fetch eth2 config when needed

paulhauner committed:
Improve error message

paulhauner committed:
Merge branch 'unstable' into add-holesky

paulhauner committed:
Tidy

paulhauner committed:
Fix clippy lints

realbigsean commented:
FYI there was discussion on the CL call about serving this state snappy compressed in checkpointz servers

antondlr commented on common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:
additional bootnode ENR: 
`- enr:-Le4QCRIcJJgCJiK8fOwIUQp6eNz5D-rbj7R4WHtglRJRYT2eGc5fqwtPeW67Op56DbADEV8ojLe9PbdLgrQ9oUad9gChGV0aDKQqX6DZjABcAAKAAAAAAAAAIJpZIJ2NIJpcIQ5gR6Wg2lwNpAgAUHQBwEQAAAAAAAAADR-iXNlY3AyNTZrMaEDb_TxT3XRqQ4rX_9OFB1kvuwscxe8ztsHqyjkkPOBbmeDdWRwgiMohHVkcDaCI4I`

jimmygchen replied:
Is this a Lighthouse bootnode? might be worth adding it here too https://github.com/eth-clients/holesky/blob/main/custom_config_data/boot_enr.txt

paulhauner replied:
Done: https://github.com/eth-clients/holesky/pull/62

paulhauner replied:
I've also added the ENR to this PR: c62f8b159

paulhauner replied:
Eh, we need to regenerate the boot ENR because the Capella fork epoch was changed.

paulhauner replied:
I've gone and regenerated an ENR on our Holesky boot node, so need for you to do it Anton ☺️

paulhauner commented:
> FYI there was discussion on the CL call about serving this state snappy compressed in checkpointz servers

Thanks, I'm currently lobbying against that: https://discord.com/channels/595666850260713488/745077610685661265/1144388115230887996

jimmygchen commented on common/eth2_network_config/built_in_network_configs/holesky/config.yaml:
```diff
@@ -0,0 +1,118 @@
+
+# Capella
+CAPELLA_FORK_VERSION: 0x30017000
+CAPELLA_FORK_EPOCH: 10
```
Updated to 256 in https://github.com/eth-clients/holesky/pull/59

```suggestion
CAPELLA_FORK_EPOCH: 256
```

jimmygchen commented on beacon_node/src/config.rs:
```diff
@@ -468,9 +468,19 @@ pub fn get_config<E: EthSpec>(
+
+    client_config.genesis_state_url_timeout =
+        clap_utils::parse_required(cli_args, "genesis-state-url-timeout")
+            .map(Duration::from_secs)?;
```
Just a note here - It would be possible for user to only specify "checkpoint-sync-url" and "checkpoint-sync-url-timeout" and get a timeout if checkpoint sync server doesn't respond with the genesis state within `genesis_state_url_timeout`. Not sure if we should change anything though.

paulhauner replied:
Good point! I've fixed that in 78f073165

jimmygchen commented on common/eth2_config/src/lib.rs:
```diff
@@ -62,6 +71,27 @@ impl Eth2Config {
+/// Describes how a genesis state may be obtained.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum GenesisStateSource {
+    /// The genesis state for this network is not yet know.
```
```suggestion
    /// The genesis state for this network is not yet known.
```

paulhauner committed:
Add SigP boot ENR

paulhauner committed:
Add newline to end of file

paulhauner committed:
Bump Capella fork epoch

paulhauner committed:
Update common/eth2_config/src/lib.rs

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen commented on common/eth2_config/src/lib.rs:
```diff
@@ -199,7 +225,7 @@ macro_rules! define_nets {
 /// perform the final step of using `std::include_bytes` to bake the files (bytes) into the binary.
 macro_rules! define_hardcoded_nets {
-    ($(($name_ident: ident, $config_dir: tt, $genesis_is_known: ident)),+) => {
+    ($(($name_ident: ident, $config_dir: tt, $genesis_is_known: path)),+) => {
         $(
         define_archive!($name_ident, $config_dir, $genesis_is_known);
```
```suggestion
    ($(($name_ident: ident, $config_dir: tt, $genesis_state_source: path)),+) => {
        $(
        define_archive!($name_ident, $config_dir, $genesis_state_source);
```

paulhauner committed:
Update common/eth2_config/src/lib.rs

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

paulhauner committed:
Use checkpoint sync URL timeout

paulhauner committed:
Bump genesis state timeout to eq checkpoint sync

paulhauner commented:
FYI I just bumped the default `--genesis-state-url-timeout` to match the existing `--checkpoint-sync-url-timeout` (180s). I don't really see why they should be different.

See 766a1c360

paulhauner commented:
I've labelled this PR as `blocked` since we need to regenerate the our ENR to suit the Capella epoch change (https://github.com/sigp/lighthouse/pull/4653/commits/1a0fccf7791eda40c671b1c64060d5aac9ebf1ff).

This branch will now generate the new ENR. I'm not exactly sure if we need to delete the old `network` directory to regenerate the key.. that might be safest.

jimmygchen commented on common/eth2_network_config/src/lib.rs:
```diff
@@ -207,20 +282,105 @@ impl Eth2NetworkConfig {
+                GenesisStateSource::Unknown
+            };
+            (state, genesis_state_source)
         } else {
```
It could be a nice feature to support `GenesisStateSource::Url` for custom testnets in the future - but may be an overkill since users have to download some other files anyway.

paulhauner replied:
Yeah it would be cool. We'd need to add a new file to define the URL, so I thought I'd leave that for future works.

jimmygchen commented on lighthouse/src/main.rs:
```diff
@@ -324,6 +324,30 @@ fn main() {
+                .help(
+                    "A URL of a beacon-API compatible server from which to download the genesis state. \
+                    Checkpoint sync server URLs can generally be used with this flag. \
+                    If not supplied, a default URL may be used. If the genesis state is already included \
```
Perhaps worth mentioning that checkpoint-sync-url will be used if this is not supplied?

paulhauner replied:
Thanks, fixed in d18edb7c5

jimmygchen commented on beacon_node/src/config.rs:
```diff
@@ -468,9 +468,30 @@ pub fn get_config<E: EthSpec>(
+    // If the `--genesis-state-url` is defined, use that to download the
+    // genesis state bytes. If it's not defined, try `--checkpoint-sync-url`.
+    client_config.genesis_state_url = if let Some(genesis_state_url) = genesis_state_url_opt {
+        Some(genesis_state_url)
```
Does `genesis_state_url` actually get used if `--checkpoint-sync-url` is supplied? Wouldn't it just go with checkpoint sync because we get `ClientGenesis::CheckpointSyncUrl` instead of `ClientGenesis::GenesisState` below?

https://github.com/sigp/lighthouse/blob/766a1c360d25698682d9f59e5c448ace300b7078/beacon_node/src/config.rs#L519-L526

paulhauner replied:
We still fetch the genesis state when doing a checkpoint sync: https://github.com/paulhauner/lighthouse/blob/d18edb7c51ec03e903dbff4148255bdddd6806e0/beacon_node/client/src/builder.rs#L380

I have a note in the PR description that talks about this too:

> When we do a fresh checkpoint sync we're downloading the genesis state to check the genesis_validators_root against the finalised state we receive. This is not entirely pointless, since we verify the checksum when we download the genesis state so we are actually guaranteeing that the finalised state is on the same network. There might be a smarter/less-download-y way to go about this, but I've run out of cycles to figure that out. Perhaps we can grab it in the next release?

jimmygchen replied:
Ah I see, I missed that. Thanks!

paulhauner committed:
Update help flag

paulhauner committed:
Update SigP boot ENR and add Teku

paulhauner commented:
I've updated our ENR to be Capella-256-ready. I also added [Teku's ENRs](https://github.com/eth-clients/holesky/pull/63/files).

See https://github.com/sigp/lighthouse/pull/4653/commits/a944a7e97bd38f788026476e34df033a21608663

paulhauner committed:
Remove duplicate ENR

paulhauner commented:
Alrighty, this one is now ready to go.

michaelsproul commented on account_manager/src/validator/slashing_protection.rs:
```diff
@@ -82,19 +83,25 @@ pub fn cli_run<T: EthSpec>(
+            context.log(),
+        )?
+        .ok_or_else(|| "Unable to get genesis state, has genesis occurred?".to_string())?
+        .genesis_validators_root();
```
It might be good (perhaps when removing Goerli's genesis) to embed the `genesis_validators_root` in the network config so we don't need to load it in cases like this. The GVR is often used for differentiating networks (including during signing), so it might be handy to have it.

paulhauner replied:
Good call, I added that in [add8bd6](https://github.com/sigp/lighthouse/pull/4653/commits/add8bd6f43a08dddcd49a2e34656db2a127f2e45)

michaelsproul commented on common/eth2_network_config/built_in_network_configs/holesky/config.yaml:
```diff
@@ -0,0 +1,118 @@
+# ---------------------------------------------------------------
+# `2**14` (= 16,384)
+MIN_GENESIS_ACTIVE_VALIDATOR_COUNT: 16384
+# Sep-15-2023 14:55:00 +UTC
+# This is an invalid valid and should be updated when you create the genesis
```
```suggestion
# Sep-15-2023 13:55:00 +UTC
```

Update this assuming my PR upstream gets merged: https://github.com/eth-clients/holesky/pull/66

michaelsproul commented on lighthouse/src/main.rs:
```diff
@@ -324,6 +324,30 @@ fn main() {
                 .global(true)
         )
+        .arg(
+            Arg::with_name("genesis-state-url")
```
There are no CLI tests for these new flags

paulhauner replied:
Good catch, I added some here: https://github.com/sigp/lighthouse/pull/4653/commits/9bdc2433320d02ce8f8db34fcd9dfbfddd3d7f0c

I wanted to test the behaviour with `--checkpoint-sync-url` however our test don't support it (they try to download a checkpoint state). I don't have the time to go down that rabbit hole at present.

michaelsproul reviewed (approved):
Looks good to me! Just a few small nitpicks, then I think we're ready to merge :rocket:

jimmygchen commented on common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml:
```diff
@@ -0,0 +1,8 @@
+# EF
+- enr:-Iq4QJk4WqRkjsX5c2CXtOra6HnxN-BMXnWhmhEQO9Bn9iABTJGdjUOurM7Btj1ouKaFkvTRoju5vz2GPmVON2dffQKGAX53x8JigmlkgnY0gmlwhLKAlv6Jc2VjcDI1NmsxoQK6S-Cii_KmfFdUJL2TANL3ksaKUnNXvTCv1tLwXs0QgIN1ZHCCIyk
+- enr:-KG4QMH842KsJOZAHxI98VJcf8oPr1U8Ylyp2Tb-sNAPniWSCaxIS4F9gc3lGOnROEok7g5qrOm8WgJTl2WXx8MhMmIMhGV0aDKQqX6DZjABcAAKAAAAAAAAAIJpZIJ2NIJpcISygIjpiXNlY3AyNTZrMaECvQMvoDF46BfJgvAbbv1hwpNu9VQBXRIpHS_B8zmkZmmDdGNwgiMog3VkcIIjKA
+- enr:-Ly4QDU8tZeygxz1gEeAD4EKe4H_8gg-IanpTY6h8A1YGPv5BPNvCMD77zjHUk_iF1pfG_8DC6jYWbIOD1k5kF-LaG4Bh2F0dG5ldHOIAAAAAAAAAACEZXRoMpCpfoNmMAFwAAoAAAAAAAAAgmlkgnY0gmlwhJK-DYCJc2VjcDI1NmsxoQN4bUae9DwIcq_56DNztksQYXeddTDKRonI5qI3YhN4SohzeW5jbmV0cwCDdGNwgiMog3VkcIIjKA
```
It looks like these two EF ENRs don't exactly match the ones on the holesky repo, has something changed?:

https://github.com/eth-clients/holesky/blob/76057d57ab1f585519ecb606a9e5f7780e925a37/custom_config_data/boot_enr.txt#L2-L3

```suggestion
- enr:-KG4QF6d6vMSboSujAXTI4vYqArccm0eIlXfcxf2Lx_VE1q6IkQo_2D5LAO3ZSBVUs0w5rrVDmABJZuMzISe_pZundADhGV0aDKQqX6DZjABcAAAAQAAAAAAAIJpZIJ2NIJpcISygIjpiXNlY3AyNTZrMaEDF3aSa7QSCvdqLpANNd8GML4PLEZVg45fKQwMWhDZjd2DdGNwgiMog3VkcIIjKA
- enr:-Ly4QJLXSSAj3ggPBIcodvBU6IyfpU_yW7E9J-5syoJorBuvcYj_Fokcjr303bQoTdWXADf8po0ssh75Mr5wVGzZZsMBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpCpfoNmMAFwAAABAAAAAAAAgmlkgnY0gmlwhJK-DYCJc2VjcDI1NmsxoQJrIlXIQDvQ6t9yDySqJYDXgZgLXzTvq8W7OI51jfmxJohzeW5jbmV0cwCDdGNwgiMog3VkcIIjKA
```

paulhauner replied:
Oh nice catch, that's probably the Capella fork update.

paulhauner committed:
Update common/eth2_network_config/built_in_network_configs/holesky/boot_enr.yaml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

paulhauner committed:
Update common/eth2_network_config/built_in_network_configs/holesky/config.yaml

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Add tests for new flags

paulhauner committed:
Add GVR to testnet config

paulhauner committed:
Fix comment

michaelsproul reviewed (approved):
Happy to merge once merge conflicts are resolved!

paulhauner committed:
Merge branch 'unstable' into add-holesky

paulhauner commented:
> Happy to merge once merge conflicts are resolved!

Thanks, I'll see how CI goes and then merge at will. Thanks for the review @jimmygchen and @michaelsproul 🙏

paulhauner committed:
Run cargo fmt

paulhauner commented:
bors r+
